plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js

```diff
@@ -0,0 +1,485 @@
+			}
+			// Return a promise that will resolve when the function is eventually called
+			return new Promise( ( resolve ) => {
+				// Store the resolve function to be called when the function executes
+				lastArgs._resolve = resolve;
+			} );
+		}
+		debounced.cancel = cancel;
+		debounced.flush = flush;
+		return debounced;
+	}
+
+	Object.entries( a8cAddressAutocompleteServiceKeys ).forEach(
+		( [ key, value ] ) => {
+			let sessionId =
+				crypto && crypto.randomUUID
+					? crypto.randomUUID()
+					: Math.random().toString( 36 ).substring( 2 );
+			let requestDurations = [];
+			let serviceErrorRetries = 0;
+			// Cache for search results - key: `${inputValue}:${country}`, value: data
+			const searchCache = new Map();
+
+			// Helper function to check cache
+			const getCachedResult = ( inputValue, country ) => {
+				const cacheKey = `${ inputValue }:${ country }`;
+				return searchCache.get( cacheKey ) || null;
+			};
+
+			// Helper function to store result in cache
+			const cacheResult = ( inputValue, country, data ) => {
+				const cacheKey = `${ inputValue }:${ country }`;
+				searchCache.set( cacheKey, data );
+
+				// Clean up cache if it gets too large (more than 100 entries)
+				if ( searchCache.size > 100 ) {
+					// Remove oldest entries (first 20 entries)
+					const entries = Array.from( searchCache.entries() );
+					entries.slice( 0, 20 ).forEach( ( [ key ] ) => {
+						searchCache.delete( key );
+					} );
+				}
+			};
+
+			const debouncedSearch = debounce(
+				async ( inputValue, country ) => {
+					const params = new URLSearchParams( {
+						query: inputValue,
+						country,
+						lang: document.documentElement.lang,
+						session_id: sessionId,
+						token: value,
+					} );
+
+					try {
+						const startTime = Date.now();
+						const response = await fetch(
+							`${ searchUrl }?${ params.toString() }`
+						);
+						const endTime = Date.now();
+						requestDurations.push( endTime - startTime );
+						let data = await response.json();
+
+						if ( data.code ) {
+							switch ( data.code ) {
+								case 'expired_jwt_token':
+								case 'malformed_jwt_token':
+								case 'invalid_jwt_token':
+								case 'invalid_issuer':
+								case 'invalid_service':
+								case 'missing_jwt_token':
+									permanentlyDisabledServices.push( key );
+									console.error(
+										'Automattic Address Suggestion has been disabled due to invalid JWT token'
+									);
+									return [];
+								case 'rate_limit_exceeded':
+									permanentlyDisabledServices.push( key );
+									setTimeout( () => {
+										permanentlyDisabledServices.splice(
+											permanentlyDisabledServices.indexOf(
+												key
+											),
+											1
+										);
+									}, response.headers.get( 'RateLimit-Retry-After' ) * 1000 );
+									console.error(
+										'Automattic Address Suggestion has been disabled due to rate limit exceeded'
+									);
+									return [];
```

Comment:
This error handling is more or less the same between this and select, except what to return and a couple of unique errors per endpoint, maybe there's a chance to move it later.